src/main/java/org/jenkinsci/plugins/workflow/support/storage/SimpleXStreamFlowNodeStorage.java

```diff
@@ -148,23 +148,71 @@ private Tag(@Nonnull FlowNode node, @Nonnull List<Action> actions) {
     private static final Field FlowNode$exec;
     private static final Field FlowNode$parents;
     private static final Field FlowNode$parentIds;
+    private static final Field FlowNode$id;
 
     static {
         XSTREAM.registerConverter(new Converter() {
             private final RobustReflectionConverter ref = new RobustReflectionConverter(XSTREAM.getMapper(), JVM.newReflectionProvider());
             // IdentityHashMap could leak memory. WeakHashMap compares by equals, which will fail with NPE in FlowNode.hashCode.
             private final Map<FlowNode,String> ids = CacheBuilder.newBuilder().weakKeys().<FlowNode,String>build().asMap();
+
+            private final Object NO_FIELD = new Object();
+
+            // Used to determine if we have a descriptorId field and if so, deduplicate them
+            private final Map<Class, Object> descriptorIdMap = CacheBuilder.newBuilder().weakKeys().<Class,Object>build().asMap();
+
             @Override public boolean canConvert(Class type) {
                 return FlowNode.class.isAssignableFrom(type);
             }
             @Override public void marshal(Object source, HierarchicalStreamWriter writer, MarshallingContext context) {
                 ref.marshal(source, writer, context);
             }
+
+            // Reflectively deduplicate the descriptorId field on flownodes
+            private void internDescriptorIdIfPresent(@Nonnull FlowNode f) {
+                try {
+                    Class c = f.getClass();
+                    Object ob = descriptorIdMap.get(c);
+                    if (ob == NO_FIELD) {
+                        return;
+                    } else if (ob != null) {
+                        Field descriptorIdField = (Field)ob;
+                        Object fieldVal = descriptorIdField.get(f);
+                        if (fieldVal != null) {
+                            descriptorIdField.set(f, ((String)fieldVal).intern());
+                        }
+                    } else { // First time we saw this class
+                        try {
+                            Field field = c.getDeclaredField("descriptorId");
```

Comment:
Way too much work. Just amend https://github.com/jenkinsci/workflow-cps-plugin/pull/82 to give `Step{Atom,Start}Node` a `readResolve`.